packages/flutter/lib/src/gestures/converter.dart

```diff
@@ -337,21 +348,44 @@ class PointerEventConverter {
                 radiusMax: radiusMax,
                 orientation: datum.orientation,
                 tilt: datum.tilt,
-              );
+              ));
             }
             break;
           case ui.PointerChange.remove:
             assert(_pointers.containsKey(datum.device));
             final _PointerState state = _pointers[datum.device];
             if (state.down) {
-              yield PointerCancelEvent(
+              yield state.record(PointerCancelEvent(
                 timeStamp: timeStamp,
                 pointer: state.pointer,
                 kind: kind,
                 device: datum.device,
+                position: state.lastPosition, // Change position in Hover
+                buttons: datum.buttons,
+                obscured: datum.obscured,
+                pressureMin: datum.pressureMin,
+                pressureMax: datum.pressureMax,
+                distance: datum.distance,
+                distanceMax: datum.distanceMax,
+                size: datum.size,
+                radiusMajor: radiusMajor,
+                radiusMinor: radiusMinor,
+                radiusMin: radiusMin,
+                radiusMax: radiusMax,
+                orientation: datum.orientation,
+                tilt: datum.tilt,
+              ));
+            }
+            if (position != state.lastPosition) {
```

Comment:
Is this some that can be fix on the embedder side? The project I am working on is moving all this expanding to embedder. The fix here will eventually move to embedder.

Reply:
Which devices that will hit this edge case?

Reply:
Yes, most of the changes in this PR are supposed to be moved to the engine (embedder?)  eventually, so that the framework will remain 1-to-1 event mapping.
I haven't met devices that will hit this edge case, but missing of this breaks the invariant of "lastEvent.position + thisEvent.delta = thisEvent.position".

Reply:
Is it only affecting the test? also, I think there are two possible possible expand, a move-> cancel-> remove
or a cancel-> hover-> remove. not sure which one is the correct intention of the remove

Reply:
You're right. To be more precise, the first one is 
```
((move -> cancel) | (hover)) -> remove
```
while the second one is 
```
cancel -> hover -> remove
```
The reason I choose the 2nd one over the 1st is complete due to complexity.

Reply:
If it is only the test is failing due to the new implementation of more secure state check, we should fix the test to remove at the right position instead. If no real device is hitting this edge case, the test is testing some that are unrealistic.

Reply:
I really don't understand the question here. Can you elaborate?

Reply:
Yes. This change adds an additional `PointerHoverEvent` when a `PointerChange.remove` features a different position than before, because currently neither of the two events Flutter expand it to allows non-zero `delta`, while `position` is updated eventually.
We're talking about whether we should do this change:
1. I think it should because it's a bug that broke the contract of "lastEvent.position + thisEvent.delta = thisEvent.position".
2. Chunheng thinks it shouldn't because no devices has hit this edge case yet.

Reply:
We don't check for position when we receive pointer remove event. This PR implement the check, and that cause one test to fail. There are two way to fix this.
1, expand the remove to hover and remove to ensure the position invariance
2, fix the test to remove at the right position.

Reply:
yes, we should send a hover if the remove event happens at a different position than the last event.

that said, that normalization should happen in the platform-specific code (java, objc, whatever), not in the dart code. the assert should be in the dart code though, if possible, so that it verifies every platform is sending sane data.

Reply:
Yes. The expansion will be moved to embedder in a PR by @chunhtai that is currently pending.